src/utils.js

```diff
@@ -0,0 +1,54 @@
+/*
+ * utils.js - utility functions to support the other code
+ *
+ * Copyright © 2022-2023 JEDLSoft
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/** Replace whitespace in input string with visible representations
+ *
+ * The following explicit mapping is used:
+ * | Whitespace | Description | Representation | Description |
+ * | - | - | - | - |
+ * | `\u0020` | regular space | `⎵` | open box |
+ * | `\u00a0` | non-breaking space | `⍽` | shouldered open box |
+ * | `\t` | tabulator | `→` | tab symbol |
+ * | `\r` | carriage return | `␍` | CR symbol |
+ * | `\n` | line feed | `␊` | LF symbol |
+ * | `\v` | vertical tab | `␋` | VT symbol |
+ *
+ * Additionally, whitespaces not included in explicit mapping are represented
+ * as their Unicode codepoint value, e.g. `\u3000` becomes `[U+3000]`
+ */
+export const withVisibleWhitespace = (/** @type {string} */ str) => {
+    // partially per https://en.wikipedia.org/wiki/Whitespace_character#Substitute_images
+    const explicitMapping = {
+        "\u0020": "⎵",
+        "\u00a0": "⍽",
+        "\t": "→",
+        "\r": "␍",
+        "\n": "␊",
+        "\v": "␋",
+    };
+    return str.replace(/\s/gm, (match) => {
```

Comment:
Maybe use `/\s/gum` so that it captures Unicode characters properly too

Reply:
Does it matter here? Per [this table](https://en.wikipedia.org/wiki/Whitespace_character#Unicode) highest codepoint for a whitespace character currently is `U+3000` and processing with a non-Unicode regex does not seem to break supplementary Unicode chars in output
<img width="319" alt="image" src="https://user-images.githubusercontent.com/15056319/233033308-856cf0aa-d0f0-4e47-b068-583039e1bb94.png">